cli/tests/e2e.rs

```diff
@@ -0,0 +1,219 @@
+use simperby_network::Peer;
+use simperby_repository::raw::RawRepository;
+use simperby_test_suite::*;
+
+fn setup_network(
+    start_fi: &FinalizationInfo,
+    server_public_key: PublicKey,
+) -> (Vec<Peer>, ServerConfig) {
+    let server_config = ServerConfig {
+        governance_port: dispense_port(),
+        consensus_port: dispense_port(),
+        repository_port: dispense_port(),
+        broadcast_interval_ms: Some(500),
+        fetch_interval_ms: Some(500),
+    };
+    let peer = vec![Peer {
+        public_key: server_public_key,
+        name: "server".to_owned(),
+        address: "127.0.0.1:1".parse().unwrap(),
+        ports: vec![
+            (
+                format!("dms-governance-{}", start_fi.header.to_hash256()),
+                server_config.governance_port,
+            ),
+            (
+                format!("dms-consensus-{}", start_fi.header.to_hash256()),
+                server_config.consensus_port,
+            ),
+            ("repository".to_owned(), server_config.repository_port),
+        ]
+        .into_iter()
+        .collect(),
+        message: "".to_owned(),
+        recently_seen_timestamp: 0,
+    }];
+    (peer, server_config)
+}
+
+fn build_simperby_cli() -> String {
+    let mut cmd = std::process::Command::new("cargo");
+    cmd.arg("build");
+    cmd.arg("--bin");
+    cmd.arg("simperby-cli");
+    cmd.arg("--release");
+    cmd.current_dir(concat!(env!("CARGO_MANIFEST_DIR"), "/../"));
+    let output = cmd.output().unwrap();
+    assert!(output.status.success());
+
+    format!(
+        "{}/../target/release/simperby-cli",
+        env!("CARGO_MANIFEST_DIR").replace('\\', "/")
+    )
+}
+
+async fn sync_each_other(cli_path: String, clients_path: Vec<String>) {
+    for path in clients_path.clone() {
+        run_command(format!("{cli_path} {path} broadcast")).await;
+    }
+    sleep_ms(200).await;
+    for path in clients_path {
+        run_command(format!("{cli_path} {path} update")).await;
+    }
+    sleep_ms(200).await;
+}
+
+#[tokio::test]
+async fn cli() {
+    setup_test();
+    let (fi, keys) = test_utils::generate_fi(4);
+    let (peers, server_config) = setup_network(&fi, keys[3].0.clone());
+    let port = server_config.repository_port;
+
+    // Setup repository and server.
+    let server_dir = create_temp_dir();
+    setup_pre_genesis_repository(&server_dir, fi.reserved_state.clone()).await;
+    // Add push configs to server repository.
+    run_command(format!(
+        "cd {server_dir} && git config receive.advertisePushOptions true"
+    ))
+    .await;
+    run_command(format!(
+        "cd {server_dir} && git config sendpack.sideband false"
+    ))
+    .await;
+
+    let cli_path = build_simperby_cli();
+    run_command(format!("{cli_path} {server_dir} genesis")).await;
+    run_command(format!("{cli_path} {server_dir} init")).await;
+
+    // Setup clients.
+    let mut clients_path = Vec::new();
+    for (_, key) in keys.iter().take(3) {
```

Comment:
Why only take 3?

Reply:
There are 3 clients and the last one is server's.

Reply:
I left a comment in Discord that it might be misleading. Let's discuss more there and leave it like this just for now.